diff --git a/scripts/northrend/crusaders_coliseum/trial_of_the_crusader/boss_jaraxxus.cpp b/scripts/northrend/crusaders_coliseum/trial_of_the_crusader/boss_jaraxxus.cpp
index dfc6fae..3d0fb4d 100644
--- a/scripts/northrend/crusaders_coliseum/trial_of_the_crusader/boss_jaraxxus.cpp
+++ b/scripts/northrend/crusaders_coliseum/trial_of_the_crusader/boss_jaraxxus.cpp
@@ -363,24 +363,24 @@ struct MANGOS_DLL_DECL boss_jaraxxusAI : public ScriptedAI
     void UpdateAI(const uint32 uiDiff)
     {
         //Return since we have no target
-		if (!m_creature->SelectHostileTarget() || !m_creature->getVictim())
-			return;
-
-		if(SpellAuraHolder* netherAura = m_creature->GetSpellAuraHolder(SPELL_NETHER_POWER))
-		{
-			if(netherAura->GetStackAmount() < m_uiMaxNetherPower && !m_bHasAura)
-			{
-				m_bHasAura = true;
-				netherAura->SetStackAmount(m_uiMaxNetherPower);
-			}
-		}
+        if (!m_creature->SelectHostileTarget() || !m_creature->getVictim())
+            return;
 
-		// temp for the old aura system
-		//if(m_creature->HasAura(SPELL_NETHER_POWER, EFFECT_INDEX_0) && !m_bHasAura)
-		//{
-		//	m_creature->GetAura(SPELL_NETHER_POWER, EFFECT_INDEX_0)->SetStackAmount(m_uiMaxNetherPower);
-		//	m_bHasAura = true;
-		//}
+        //if(SpellAuraHolder* netherAura = m_creature->GetSpellAuraHolder(SPELL_NETHER_POWER))
+        //{
+        //    if(netherAura->GetStackAmount() < m_uiMaxNetherPower && !m_bHasAura)
+        //    {
+        //        m_bHasAura = true;
+        //        netherAura->SetStackAmount(m_uiMaxNetherPower);
+        //    }
+        //}
+
+        // temp for the old aura system
+        if(m_creature->HasAura(SPELL_NETHER_POWER, EFFECT_INDEX_0) && !m_bHasAura)
+        {
+            m_creature->GetAura(SPELL_NETHER_POWER, EFFECT_INDEX_0)->SetStackAmount(m_uiMaxNetherPower);
+            m_bHasAura = true;
+        }
 
         // spells
         if (m_uiIncinerateFleshTimer < uiDiff)
diff --git a/scripts/northrend/ulduar/ulduar/boss_auriaya.cpp b/scripts/northrend/ulduar/ulduar/boss_auriaya.cpp
index c6def78..33ddb7a 100644
--- a/scripts/northrend/ulduar/ulduar/boss_auriaya.cpp
+++ b/scripts/northrend/ulduar/ulduar/boss_auriaya.cpp
@@ -257,15 +257,15 @@ struct MANGOS_DLL_DECL mob_feral_defenderAI : public ScriptedAI
             m_creature->SetFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NON_ATTACKABLE);
             if (m_creature->HasAura(SPELL_FERAL_ESSENCE))
             {
-				if(SpellAuraHolder* strenght = m_creature->GetSpellAuraHolder(SPELL_FERAL_ESSENCE))
-				{
-					if(strenght->ModStackAmount(-1))
-						m_creature->RemoveAurasDueToSpell(SPELL_FERAL_ESSENCE);
-				}
+                //if(SpellAuraHolder* strenght = m_creature->GetSpellAuraHolder(SPELL_FERAL_ESSENCE))
+                //{
+                //    if(strenght->ModStackAmount(-1))
+                //        m_creature->RemoveAurasDueToSpell(SPELL_FERAL_ESSENCE);
+                //}
 
-				// temp for old aura system
-				//if(m_creature->GetAura(SPELL_FERAL_ESSENCE, EFFECT_INDEX_0)->modStackAmount(-1))	 	
-				//	m_creature->RemoveAurasDueToSpell(SPELL_FERAL_ESSENCE);
+                // temp for old aura system
+                if(m_creature->GetAura(SPELL_FERAL_ESSENCE, EFFECT_INDEX_0)->modStackAmount(-1))	 	
+                    m_creature->RemoveAurasDueToSpell(SPELL_FERAL_ESSENCE);
 
                 m_uiRevive_Delay = 35000;
                 m_bIsDead = true;
@@ -281,23 +281,23 @@ struct MANGOS_DLL_DECL mob_feral_defenderAI : public ScriptedAI
         if (!m_creature->SelectHostileTarget() || !m_creature->getVictim())
             return;
 
-		if(SpellAuraHolder* essence = m_creature->GetSpellAuraHolder(SPELL_FERAL_ESSENCE))
-		{
-			if(essence->GetStackAmount() < 9 && !m_bHasAura)
-			{
-				m_bHasAura = true;
-				essence->SetStackAmount(9);
-			}
-		}
-
-		// temp for the old aura system
-		//if(m_creature->HasAura(SPELL_FERAL_ESSENCE, EFFECT_INDEX_0) && !m_bHasAura)
-		//{
-		//	m_creature->GetAura(SPELL_FERAL_ESSENCE, EFFECT_INDEX_0)->SetStackAmount(9);
-		//	m_bHasAura = true;
-		//}
-
-		if (m_uiPounce_Timer < diff)
+        //if(SpellAuraHolder* essence = m_creature->GetSpellAuraHolder(SPELL_FERAL_ESSENCE))
+        //{
+        //    if(essence->GetStackAmount() < 9 && !m_bHasAura)
+        //    {
+        //        m_bHasAura = true;
+        //        essence->SetStackAmount(9);
+        //    }
+        //}
+
+        // temp for the old aura system
+        if(m_creature->HasAura(SPELL_FERAL_ESSENCE, EFFECT_INDEX_0) && !m_bHasAura)
+        {
+            m_creature->GetAura(SPELL_FERAL_ESSENCE, EFFECT_INDEX_0)->SetStackAmount(9);
+            m_bHasAura = true;
+        }
+
+        if (m_uiPounce_Timer < diff)
         {
             if (Unit* target = m_creature->SelectAttackingTarget(ATTACKING_TARGET_RANDOM, 0))
             {
diff --git a/scripts/northrend/ulduar/ulduar/boss_freya.cpp b/scripts/northrend/ulduar/ulduar/boss_freya.cpp
index c0d4742..3cfe73a 100644
--- a/scripts/northrend/ulduar/ulduar/boss_freya.cpp
+++ b/scripts/northrend/ulduar/ulduar/boss_freya.cpp
@@ -858,22 +858,22 @@ struct MANGOS_DLL_DECL boss_freyaAI : public ScriptedAI
             if (!m_creature->SelectHostileTarget() || !m_creature->getVictim())
                 return;
 
-            if(SpellAuraHolder* natureAura = m_creature->GetSpellAuraHolder(SPELL_ATTUNED_TO_NATURE))
-            {
-                if(natureAura->GetStackAmount() < 150 && !m_bHasAura)
-                {
-                    m_bHasAura = true;
-                    natureAura->SetStackAmount(150);
-                }
-            }
-
-            // temp for the old aura system
-            //if(m_creature->HasAura(SPELL_ATTUNED_TO_NATURE, EFFECT_INDEX_0) && !m_bHasAura)
+            //if(SpellAuraHolder* natureAura = m_creature->GetSpellAuraHolder(SPELL_ATTUNED_TO_NATURE))
             //{
-            //	m_creature->GetAura(SPELL_ATTUNED_TO_NATURE, EFFECT_INDEX_0)->SetStackAmount(150);
-            //	m_bHasAura = true;
+            //    if(natureAura->GetStackAmount() < 150 && !m_bHasAura)
+            //    {
+            //        m_bHasAura = true;
+            //        natureAura->SetStackAmount(150);
+            //    }
             //}
 
+            // temp for the old aura system
+            if(m_creature->HasAura(SPELL_ATTUNED_TO_NATURE, EFFECT_INDEX_0) && !m_bHasAura)
+            {
+                m_creature->GetAura(SPELL_ATTUNED_TO_NATURE, EFFECT_INDEX_0)->SetStackAmount(150);
+                m_bHasAura = true;
+            }
+
             if(!m_creature->HasAura(m_bIsRegularMode ? SPELL_TOUCH_OF_EONAR : SPELL_TOUCH_OF_EONAR_H))
                 DoCast(m_creature, m_bIsRegularMode ? SPELL_TOUCH_OF_EONAR : SPELL_TOUCH_OF_EONAR_H);
 
@@ -909,15 +909,15 @@ struct MANGOS_DLL_DECL boss_freyaAI : public ScriptedAI
                     if(!pWaterSpirit->isAlive() && !pStormLasher->isAlive() && !pSnapLasher->isAlive())
                     {
                         m_bWaveDeadCheck = false;
-                        if(SpellAuraHolder* natureAura = m_creature->GetSpellAuraHolder(SPELL_ATTUNED_TO_NATURE))
-                        {
-                            if(natureAura->ModStackAmount(-30))
-                                m_creature->RemoveAurasDueToSpell(SPELL_ATTUNED_TO_NATURE);
-                        }
+                        //if(SpellAuraHolder* natureAura = m_creature->GetSpellAuraHolder(SPELL_ATTUNED_TO_NATURE))
+                        //{
+                        //    if(natureAura->ModStackAmount(-30))
+                        //        m_creature->RemoveAurasDueToSpell(SPELL_ATTUNED_TO_NATURE);
+                        //}
 
                         // temp for old aura system
-                        //if(m_creature->GetAura(SPELL_ATTUNED_TO_NATURE, EFFECT_INDEX_0)->modStackAmount(-30))	 	
-                        //	m_creature->RemoveAurasDueToSpell(SPELL_ATTUNED_TO_NATURE);
+                        if(m_creature->GetAura(SPELL_ATTUNED_TO_NATURE, EFFECT_INDEX_0)->modStackAmount(-30))	 	
+                            m_creature->RemoveAurasDueToSpell(SPELL_ATTUNED_TO_NATURE);
                     }
                     else
                     {
@@ -1345,7 +1345,7 @@ struct MANGOS_DLL_DECL mob_freya_spawnedAI : public ScriptedAI
         m_uiDeathCountdown          = 10000;
         m_uiTidalWave_Timer         = urand(2000,4000);
         m_uiStormbolt_Timer         = 1000;
-        m_uiLightningLash_Timer     = urand(11000,14000);        
+        m_uiLightningLash_Timer     = urand(11000,14000);
         m_uiFlameLash_Timer         = urand(5000,10000);
         m_uiNaturesFury_Timer       = urand(8000,10000);
         m_uiRespawnSpores_Timer     = 5000;
@@ -1381,15 +1381,15 @@ struct MANGOS_DLL_DECL mob_freya_spawnedAI : public ScriptedAI
         {
             if (Creature* pFreya = m_creature->GetMap()->GetCreature(m_pInstance->GetData64(NPC_FREYA)))
             {
-                if(SpellAuraHolder* natureAura = pFreya->GetSpellAuraHolder(SPELL_ATTUNED_TO_NATURE))
-                {
-                    if(natureAura->ModStackAmount(-25))
-                        m_creature->RemoveAurasDueToSpell(SPELL_ATTUNED_TO_NATURE);
-                }
+                //if(SpellAuraHolder* natureAura = pFreya->GetSpellAuraHolder(SPELL_ATTUNED_TO_NATURE))
+                //{
+                //    if(natureAura->ModStackAmount(-25))
+                //        m_creature->RemoveAurasDueToSpell(SPELL_ATTUNED_TO_NATURE);
+                //}
 
                 // temp for old aura system
-                //if(pFreya->GetAura(SPELL_ATTUNED_TO_NATURE, EFFECT_INDEX_0)->modStackAmount(-25))	 	
-                //	pFreya->RemoveAurasDueToSpell(SPELL_ATTUNED_TO_NATURE);
+                if(pFreya->GetAura(SPELL_ATTUNED_TO_NATURE, EFFECT_INDEX_0)->modStackAmount(-25))	 	
+                    pFreya->RemoveAurasDueToSpell(SPELL_ATTUNED_TO_NATURE);
             }
         }
 
@@ -1397,15 +1397,15 @@ struct MANGOS_DLL_DECL mob_freya_spawnedAI : public ScriptedAI
         {
             if (Creature* pFreya = m_creature->GetMap()->GetCreature(m_pInstance->GetData64(NPC_FREYA)))
             {
-                if(SpellAuraHolder* natureAura = pFreya->GetSpellAuraHolder(SPELL_ATTUNED_TO_NATURE))
-                {
-                    if(natureAura->ModStackAmount(-2))
-                        m_creature->RemoveAurasDueToSpell(SPELL_ATTUNED_TO_NATURE);
-                }
+                //if(SpellAuraHolder* natureAura = pFreya->GetSpellAuraHolder(SPELL_ATTUNED_TO_NATURE))
+                //{
+                //    if(natureAura->ModStackAmount(-2))
+                //        m_creature->RemoveAurasDueToSpell(SPELL_ATTUNED_TO_NATURE);
+                //}
 
                 // temp for old aura system
-                //if(pFreya->GetAura(SPELL_ATTUNED_TO_NATURE, EFFECT_INDEX_0)->modStackAmount(-2))	 	
-                //	pFreya->RemoveAurasDueToSpell(SPELL_ATTUNED_TO_NATURE);
+                if(pFreya->GetAura(SPELL_ATTUNED_TO_NATURE, EFFECT_INDEX_0)->modStackAmount(-2))	 	
+                    pFreya->RemoveAurasDueToSpell(SPELL_ATTUNED_TO_NATURE);
             }
         }
     }
diff --git a/scripts/northrend/ulduar/ulduar/boss_ignis.cpp b/scripts/northrend/ulduar/ulduar/boss_ignis.cpp
index 5f6dfa3..0e9af26 100644
--- a/scripts/northrend/ulduar/ulduar/boss_ignis.cpp
+++ b/scripts/northrend/ulduar/ulduar/boss_ignis.cpp
@@ -164,15 +164,15 @@ struct MANGOS_DLL_DECL mob_iron_constructAI : public ScriptedAI
             {
                 if (pTemp->HasAura(BUFF_STRENGHT_OF_CREATOR))
                 {
-                    if(SpellAuraHolder* strenght = pTemp->GetSpellAuraHolder(BUFF_STRENGHT_OF_CREATOR))
-                    {
-                        if(strenght->ModStackAmount(-1))
-                            pTemp->RemoveAurasDueToSpell(BUFF_STRENGHT_OF_CREATOR);
-                    }
+                    //if(SpellAuraHolder* strenght = pTemp->GetSpellAuraHolder(BUFF_STRENGHT_OF_CREATOR))
+                    //{
+                    //    if(strenght->ModStackAmount(-1))
+                    //        pTemp->RemoveAurasDueToSpell(BUFF_STRENGHT_OF_CREATOR);
+                    //}
 
                     // temp for old aura system
-                    //if(pTemp->GetAura(BUFF_STRENGHT_OF_CREATOR, EFFECT_INDEX_0)->modStackAmount(-1))	 	
-                    //	pTemp->RemoveAurasDueToSpell(BUFF_STRENGHT_OF_CREATOR);
+                    if(pTemp->GetAura(BUFF_STRENGHT_OF_CREATOR, EFFECT_INDEX_0)->modStackAmount(-1))	 	
+                        pTemp->RemoveAurasDueToSpell(BUFF_STRENGHT_OF_CREATOR);
                 }
             }
         }
diff --git a/scripts/northrend/ulduar/ulduar/boss_yogg_saron.cpp b/scripts/northrend/ulduar/ulduar/boss_yogg_saron.cpp
index c51c2b8..9410043 100644
--- a/scripts/northrend/ulduar/ulduar/boss_yogg_saron.cpp
+++ b/scripts/northrend/ulduar/ulduar/boss_yogg_saron.cpp
@@ -1935,35 +1935,35 @@ struct MANGOS_DLL_DECL mob_immortal_guardianAI : public ScriptedAI
         if (!m_creature->SelectHostileTarget() || !m_creature->getVictim())
             return;
 
-        if(SpellAuraHolder* empoweredAura = m_creature->GetSpellAuraHolder(SPELL_EMPOWERED))
-        {
-            if(empoweredAura->GetStackAmount() < 9 && !m_bHasAura)
-            {
-                m_bHasAura = true;
-                empoweredAura->SetStackAmount(9);
-            }
-        }
-
-        // temp for the old aura system
-        //if(m_creature->HasAura(SPELL_EMPOWERED, EFFECT_INDEX_0) && !m_bHasAura)
+        //if(SpellAuraHolder* empoweredAura = m_creature->GetSpellAuraHolder(SPELL_EMPOWERED))
         //{
-        //	m_creature->GetAura(SPELL_EMPOWERED, EFFECT_INDEX_0)->SetStackAmount(9);
-        //	m_bHasAura = true;
+        //    if(empoweredAura->GetStackAmount() < 9 && !m_bHasAura)
+        //    {
+        //        m_bHasAura = true;
+        //        empoweredAura->SetStackAmount(9);
+        //    }
         //}
 
+        // temp for the old aura system
+        if(m_creature->HasAura(SPELL_EMPOWERED, EFFECT_INDEX_0) && !m_bHasAura)
+        {
+            m_creature->GetAura(SPELL_EMPOWERED, EFFECT_INDEX_0)->SetStackAmount(9);
+            m_bHasAura = true;
+        }
+
         if(m_creature->GetHealthPercent() > 10)
         {
             if(m_creature->GetHealthPercent() < m_uiHealth)
             {
-                if(SpellAuraHolder* empoweredAura = m_creature->GetSpellAuraHolder(SPELL_EMPOWERED))
-                {
-                    if(empoweredAura->ModStackAmount(-1))
-                        m_creature->RemoveAurasDueToSpell(SPELL_EMPOWERED);
-                }
+                //if(SpellAuraHolder* empoweredAura = m_creature->GetSpellAuraHolder(SPELL_EMPOWERED))
+                //{
+                //    if(empoweredAura->ModStackAmount(-1))
+                //        m_creature->RemoveAurasDueToSpell(SPELL_EMPOWERED);
+                //}
 
                 // temp for old aura system
-                //if(m_creature->GetAura(SPELL_EMPOWERED, EFFECT_INDEX_0)->modStackAmount(-1))	 	
-                //	m_creature->RemoveAurasDueToSpell(SPELL_EMPOWERED);
+                if(m_creature->GetAura(SPELL_EMPOWERED, EFFECT_INDEX_0)->modStackAmount(-1))	 	
+                    m_creature->RemoveAurasDueToSpell(SPELL_EMPOWERED);
                 m_uiHealth -= 10;
             }
         }
