diff --git a/src/game/SpellMgr.cpp b/src/game/SpellMgr.cpp
index 29577ac..ac86639 100644
--- a/src/game/SpellMgr.cpp
+++ b/src/game/SpellMgr.cpp
@@ -1721,6 +1721,16 @@ bool SpellMgr::IsNoStackSpellDueToSpell(uint32 spellId_1, uint32 spellId_2) cons
             {
                 case SPELLFAMILY_GENERIC:                   // same family case
                 {
+                    // Dark Essence & Light Essence
+                    if ((spellInfo_1->Id == 65684 && spellInfo_2->Id == 65686) ||
+                        (spellInfo_2->Id == 65684 && spellInfo_1->Id == 65686))
+                        return true;
+
+					// Potent Fungus and Mini must remove each other (Amanitar encounter, Ahn'kahet)
+                    if ((spellInfo_1->Id == 57055 && spellInfo_2->Id == 56648) ||
+                        (spellInfo_2->Id == 57055 && spellInfo_1->Id == 56648))
+                        return true;
+
                     // Thunderfury
                     if ((spellInfo_1->Id == 21992 && spellInfo_2->Id == 27648) ||
                         (spellInfo_2->Id == 21992 && spellInfo_1->Id == 27648))
@@ -1769,6 +1779,19 @@ bool SpellMgr::IsNoStackSpellDueToSpell(uint32 spellId_1, uint32 spellId_2) cons
                     if( spellInfo_1->SpellIconID == 3559 && spellInfo_2->SpellIconID == 3559 )
                         return false;
 
+					// Flame Tsunami Visual and Damage
+                    if ((spellInfo_1->Id == 57492 && spellInfo_2->Id == 57494) ||
+                        (spellInfo_2->Id == 57492 && spellInfo_1->Id == 57494))
+                        return false;
+
+                    // Sextant of Unstable Currents and Band of the Eternal Sage
+                    if( spellInfo_1->SpellIconID == 502 && spellInfo_2->SpellIconID == 502 )
+                        return false;
+
+                    // Lightning Speed and Crushing Waves
+                    if( spellInfo_1->SpellIconID == 2010 && spellInfo_2->SpellIconID == 2010 )
+                        return false;
+					
                     break;
                 }
                 case SPELLFAMILY_MAGE:
@@ -1832,6 +1855,17 @@ bool SpellMgr::IsNoStackSpellDueToSpell(uint32 spellId_1, uint32 spellId_2) cons
                     // Blessing of Sanctuary (multi-family check, some from 16 spell icon spells)
                     if (spellInfo_1->Id == 67480 && spellInfo_2->Id == 20911)
                         return false;
+                    // Seal of Command and Frostforged Champion (multi-family check)
+                    if (spellInfo_1->Id == 72412 && spellInfo_2->SpellIconID == 561)
+                        return false;
+
+                    // Hardened Skin and Devotion Aura (multi-family check)
+                    if (spellInfo_1->Id == 71586 && spellInfo_2->SpellIconID == 291)
+                        return false;
+
+                    // Aegis and Holy Shield (multi-family check)
+                    if (spellInfo_1->Id == 67631 && spellInfo_2->SpellIconID == 453)
+                        return false;
 
                     break;
                 }
@@ -1839,6 +1873,11 @@ bool SpellMgr::IsNoStackSpellDueToSpell(uint32 spellId_1, uint32 spellId_2) cons
             // Dragonmaw Illusion, Blood Elf Illusion, Human Illusion, Illidari Agent Illusion, Scarlet Crusade Disguise
             if(spellInfo_1->SpellIconID == 1691 && spellInfo_2->SpellIconID == 1691)
                 return false;
+
+			// Drums of the Wild and Gift of the Wild
+			if( (spellInfo_1->SpellIconID == 2435 && spellInfo_2->SpellIconID == 123) ||
+				(spellInfo_2->SpellIconID == 2435 && spellInfo_1->SpellIconID == 123) )
+				return true;
             break;
         case SPELLFAMILY_MAGE:
             if( spellInfo_2->SpellFamilyName == SPELLFAMILY_MAGE )
@@ -1875,6 +1914,15 @@ bool SpellMgr::IsNoStackSpellDueToSpell(uint32 spellId_1, uint32 spellId_2) cons
             if( spellInfo_1->SpellIconID == 125 && spellInfo_2->Id == 18820 )
                 return false;
 
+            // Fire Ward and Fire Shield (multi-family check)
+            if ((spellInfo_1->SpellIconID == 16) && (spellInfo_1->SpellFamilyFlags & UI64LIT(0x0000000000000008)) &&
+                (spellInfo_2->SpellIconID == 16) && (spellInfo_2->SpellFamilyName == SPELLFAMILY_WARLOCK))
+                return false;
+
+			// Hot Streak and Firestarter
+            if( spellInfo_1->EffectSpellClassMaskC[0] == 262144 && spellInfo_2->EffectSpellClassMaskC[0] == 262144)
+                return false;
+
             break;
         case SPELLFAMILY_WARLOCK:
             if( spellInfo_2->SpellFamilyName == SPELLFAMILY_WARLOCK )
@@ -1903,10 +1951,20 @@ bool SpellMgr::IsNoStackSpellDueToSpell(uint32 spellId_1, uint32 spellId_2) cons
                 // Metamorphosis, diff effects
                 if (spellInfo_1->SpellIconID == 3314 && spellInfo_2->SpellIconID == 3314)
                     return false;
+
+				// Nether Protection effects
+                if( spellInfo_2->SpellIconID==1985 && spellInfo_1->SpellIconID==1985 && spellInfo_1->SpellVisual[0]==9750 )
+                    return false;
             }
             // Detect Invisibility and Mana Shield (multi-family check)
             if( spellInfo_1->Id == 132 && spellInfo_2->SpellIconID == 209 && spellInfo_2->SpellVisual[0] == 968 )
                 return false;
+
+            // Fire Shield and Fire Ward (multi-family check)
+            if ((spellInfo_1->SpellIconID == 16) &&
+                (spellInfo_2->SpellIconID == 16) && (spellInfo_2->SpellFamilyFlags & UI64LIT(0x0000000000000008)))
+                return false;
+
             break;
         case SPELLFAMILY_WARRIOR:
             if( spellInfo_2->SpellFamilyName == SPELLFAMILY_WARRIOR )
@@ -1920,8 +1978,17 @@ bool SpellMgr::IsNoStackSpellDueToSpell(uint32 spellId_1, uint32 spellId_2) cons
                 if( (spellInfo_1->SpellIconID == 456 && spellInfo_2->SpellIconID == 2006) ||
                     (spellInfo_2->SpellIconID == 456 && spellInfo_1->SpellIconID == 2006) )
                     return false;
+
+				// Taste of Blood and Sudden Death
+                if( (spellInfo_1->Id == 52437 && spellInfo_2->Id == 60503) ||
+                    (spellInfo_2->Id == 52437 && spellInfo_1->Id == 60503) )
+                    return false;
             }
 
+			// Flurry and Rip stacking
+            if(spellInfo_1->SpellIconID == 108 && spellInfo_2->SpellIconID == 108)
+                return false;
+
             // Hamstring -> Improved Hamstring (multi-family check)
             if( (spellInfo_1->SpellFamilyFlags & UI64LIT(0x2)) && spellInfo_2->Id == 23694 )
                 return false;
@@ -1956,7 +2023,12 @@ bool SpellMgr::IsNoStackSpellDueToSpell(uint32 spellId_1, uint32 spellId_2) cons
         case SPELLFAMILY_DRUID:
             if( spellInfo_2->SpellFamilyName == SPELLFAMILY_DRUID )
             {
-                //Omen of Clarity and Blood Frenzy
+                // Lacerate and Moonfire
+                 if (spellInfo_1->SpellIconID == 225 && spellInfo_2->SpellIconID == 2246 ||
+                     spellInfo_2->SpellIconID == 225 && spellInfo_1->SpellIconID == 2246 )
+                    return false;
+                
+				//Omen of Clarity and Blood Frenzy
                 if( (spellInfo_1->SpellFamilyFlags == UI64LIT(0x0) && spellInfo_1->SpellIconID == 108) && (spellInfo_2->SpellFamilyFlags & UI64LIT(0x20000000000000)) ||
                     (spellInfo_2->SpellFamilyFlags == UI64LIT(0x0) && spellInfo_2->SpellIconID == 108) && (spellInfo_1->SpellFamilyFlags & UI64LIT(0x20000000000000)) )
                     return false;
@@ -2000,6 +2072,15 @@ bool SpellMgr::IsNoStackSpellDueToSpell(uint32 spellId_1, uint32 spellId_2) cons
                     return false;
             }
 
+			// Flurry and Rip stacking
+            if(spellInfo_1->SpellIconID == 108 && spellInfo_2->SpellIconID == 108)
+                return false;
+
+			// Drums of the Wild and Gift of the Wild
+			if( (spellInfo_1->SpellIconID == 2435 && spellInfo_2->SpellIconID == 123) ||
+				(spellInfo_2->SpellIconID == 2435 && spellInfo_1->SpellIconID == 123) )
+				return true;
+
             // Leader of the Pack and Scroll of Stamina (multi-family check)
             if( spellInfo_1->Id == 24932 && spellInfo_2->SpellIconID == 312 && spellInfo_2->SpellVisual[0] == 216 )
                 return false;
@@ -2072,6 +2153,10 @@ bool SpellMgr::IsNoStackSpellDueToSpell(uint32 spellId_1, uint32 spellId_2) cons
                 if (IsSealSpell(spellInfo_1) && IsSealSpell(spellInfo_2))
                     return true;
 
+                // Repentance removes Righteous Vengeance
+                if (spellInfo_1->Id == 20066 && spellInfo_2->Id == 61840)
+                    return true;
+
                 // Swift Retribution / Improved Devotion Aura (talents) and Paladin Auras
                 if ((spellInfo_1->SpellFamilyFlags2 & 0x00000020) && (spellInfo_2->SpellIconID == 291 || spellInfo_2->SpellIconID == 3028) ||
                     (spellInfo_2->SpellFamilyFlags2 & 0x00000020) && (spellInfo_1->SpellIconID == 291 || spellInfo_1->SpellIconID == 3028))
@@ -2098,6 +2183,10 @@ bool SpellMgr::IsNoStackSpellDueToSpell(uint32 spellId_1, uint32 spellId_2) cons
             if (spellInfo_2->Id == 67480 && spellInfo_1->Id == 20911)
                 return false;
 
+            // Inner Fire and Consecration
+            if((spellInfo_2->SpellFamilyName == SPELLFAMILY_PRIEST) && (spellInfo_1->SpellIconID == 51 && spellInfo_2->SpellIconID == 51))
+                return false;
+
             // Combustion and Fire Protection Aura (multi-family check)
             if( spellInfo_2->Id == 11129 && spellInfo_1->SpellIconID == 33 && spellInfo_1->SpellVisual[0] == 321 )
                 return false;
@@ -2109,6 +2198,23 @@ bool SpellMgr::IsNoStackSpellDueToSpell(uint32 spellId_1, uint32 spellId_2) cons
             // *Seal of Command and Band of Eternal Champion (multi-family check)
             if( spellInfo_1->SpellIconID==561 && spellInfo_1->SpellVisual[0]==7992 && spellId_2 == 35081)
                 return false;
+
+			// Seal of Command and Frostforged Champion (multi-family check)
+            if( spellInfo_2->SpellFamilyName == SPELLFAMILY_GENERIC )
+            { 
+                // Seal of Command and Frostforged Champion (multi-family check) 
+                if( spellInfo_1->SpellIconID == 561 && spellInfo_2->Id == 72412 )
+                    return false;
+
+                // Hardened Skin and Devotion Aura (multi-family check)
+                if( spellInfo_1->SpellIconID == 291 && spellInfo_2->Id == 71586 )
+                    return false;
+                
+                // Aegis and Holy Shield (multi-family check)
+                if( spellInfo_1->SpellIconID == 453 && spellInfo_2->Id == 67631 )
+                    return false;
+            }
+
             break;
         case SPELLFAMILY_SHAMAN:
             if( spellInfo_2->SpellFamilyName == SPELLFAMILY_SHAMAN )
