diff --git a/src/game/SpellEffects.cpp b/src/game/SpellEffects.cpp
index cd4ad0d..0763533 100644
--- a/src/game/SpellEffects.cpp
+++ b/src/game/SpellEffects.cpp
@@ -520,7 +520,8 @@ void Spell::EffectSchoolDMG(SpellEffectIndex effect_idx)
                     if (aura)
                     {
                         int32 damagetick = aura->GetModifier()->m_amount;
-                        damage += damagetick * 4;
+	                 	damage += damagetick * 21/8;
+           				m_currentBasePoints[1]=damage * 2/9;
 
                         // Glyph of Conflagrate
                         if (!m_caster->HasAura(56235))
