diff --git a/src/game/Spell.cpp b/src/game/Spell.cpp
index 7ce9dcc..4505551 100644
--- a/src/game/Spell.cpp
+++ b/src/game/Spell.cpp
@@ -2942,7 +2942,7 @@ void Spell::cast(bool skipCheck)
     InitializeDamageMultipliers();
 
     // Okay, everything is prepared. Now we need to distinguish between immediate and evented delayed spells
-    if (m_spellInfo->speed > 0.0f)
+	if (m_spellInfo->speed > 0.0f || m_spellInfo->Id == 14157)                // hack for Ruthlessness
     {
 
         // Remove used for cast item if need (it can be already NULL after TakeReagents call
