diff --git a/src/game/Spell.cpp b/src/game/Spell.cpp
index 322ecc6..f7ef9f4 100644
--- a/src/game/Spell.cpp
+++ b/src/game/Spell.cpp
@@ -6307,6 +6307,11 @@ bool Spell::CheckTarget( Unit* target, SpellEffectIndex eff )
             return false;
     }
 
+	// Check Sated & Exhaustion debuffs
+    if (((m_spellInfo->Id == 2825) && (target->HasAura(57724))) ||
+        ((m_spellInfo->Id == 32182) && (target->HasAura(57723))))
+        return false;
+
     // Check targets for LOS visibility (except spells without range limitations )
     switch(m_spellInfo->Effect[eff])
     {
