diff --git a/src/game/UnitAuraProcHandler.cpp b/src/game/UnitAuraProcHandler.cpp
index 1415948..9a4a197 100644
--- a/src/game/UnitAuraProcHandler.cpp
+++ b/src/game/UnitAuraProcHandler.cpp
@@ -1918,6 +1918,10 @@ SpellAuraProcResult Unit::HandleDummyAuraProc(Unit *pVictim, uint32 damage, Aura
                     if (GetGUID() == triggeredByAura->GetCasterGUID())
                         return SPELL_AURA_PROC_FAILED;
 
+					// Prevent Beacon of Light from taking into consideration JoL self heal 
+                    if (procSpell->Id == 20267)
+                         return SPELL_AURA_PROC_FAILED;
+
                     // beacon
                     Unit* beacon = triggeredByAura->GetCaster();
                     if (!beacon)
