diff --git a/src/game/StatSystem.cpp b/src/game/StatSystem.cpp
index f6e30e1..c159f8e 100644
--- a/src/game/StatSystem.cpp
+++ b/src/game/StatSystem.cpp
@@ -876,7 +876,34 @@ bool Pet::UpdateStats(Stats stat)
     if ( stat == STAT_STAMINA )
     {
         if(owner)
-            value += float(owner->GetStat(stat)) * 0.3f;
+        {
+            float scale_coeff = 0.3f;
+            switch (owner->getClass())
+            {
+                case CLASS_HUNTER:
+                    scale_coeff = 0.4493f;
+                    break;
+                case CLASS_WARLOCK:
+                {
+                    CreatureInfo const *cinfo = GetCreatureInfo();
+                    CreatureFamily petFamily = (CreatureFamily) cinfo->family;
+                    switch (petFamily)
+                    {
+                        case CREATURE_FAMILY_FELHUNTER:  value += float(owner->GetStat(stat)) * 0.7125f; break;
+                        case CREATURE_FAMILY_VOIDWALKER: value += float(owner->GetStat(stat)) * 0.825f; break;
+                        case CREATURE_FAMILY_FELGUARD:   value += float(owner->GetStat(stat)) * 0.825f; break;
+                        case CREATURE_FAMILY_SUCCUBUS:   value += float(owner->GetStat(stat)) * 0.6825f; break;
+                        case CREATURE_FAMILY_IMP:        value += float(owner->GetStat(stat)) * 0.63f; break;
+                        default: value += float(owner->GetStat(stat)) * 0.3f; break;
+                    }
+                    break;
+                }
+                case CLASS_DEATH_KNIGHT:
+                    scale_coeff = 0.3928f;
+                    break;
+            }
+            value += float(owner->GetStat(stat)) * scale_coeff;
+        } 
     }
                                                             //warlock's and mage's pets gain 30% of owner's intellect
     else if ( stat == STAT_INTELLECT && getPetType() == SUMMON_PET )
