diff --git a/src/game/Spell.cpp b/src/game/Spell.cpp
index d67e8d3..8b6d0c4 100644
--- a/src/game/Spell.cpp
+++ b/src/game/Spell.cpp
@@ -5146,6 +5146,13 @@ SpellCastResult Spell::CheckCast(bool strict)
                     return SPELL_FAILED_BAD_TARGETS;
                 break;
             }
+            case SPELL_EFFECT_LEAP_BACK:
+            {
+                if(m_spellInfo->Id == 781)
+                    if(!m_caster->isInCombat()) 
+                        return SPELL_FAILED_CANT_DO_THAT_RIGHT_NOW; 
+                break;
+            }
             default:break;
         }
     }
