diff --git a/src/game/SpellAuras.cpp b/src/game/SpellAuras.cpp
index c268315..5a5b50d 100644
--- a/src/game/SpellAuras.cpp
+++ b/src/game/SpellAuras.cpp
@@ -5810,7 +5810,6 @@ void Aura::HandleShapeshiftBoosts(bool apply)
     {
         case FORM_CAT:
             spellId1 = 3025;
-            HotWSpellId = 24900;
             MasterShaperSpellId = 48420;
             break;
         case FORM_TREE:
@@ -5827,13 +5826,11 @@ void Aura::HandleShapeshiftBoosts(bool apply)
         case FORM_BEAR:
             spellId1 = 1178;
             spellId2 = 21178;
-            HotWSpellId = 24899;
             MasterShaperSpellId = 48418;
             break;
         case FORM_DIREBEAR:
             spellId1 = 9635;
             spellId2 = 21178;
-            HotWSpellId = 24899;
             MasterShaperSpellId = 48418;
             break;
         case FORM_BATTLESTANCE:
@@ -5958,7 +5955,7 @@ void Aura::HandleShapeshiftBoosts(bool apply)
             }
 
             // Heart of the Wild
-            if (HotWSpellId)
+            if (form == FORM_CAT || form == FORM_BEAR || form == FORM_DIREBEAR)
             {
                 Unit::AuraList const& mModTotalStatPct = target->GetAurasByType(SPELL_AURA_MOD_TOTAL_STAT_PERCENTAGE);
                 for(Unit::AuraList::const_iterator i = mModTotalStatPct.begin(); i != mModTotalStatPct.end(); ++i)
@@ -5966,10 +5963,35 @@ void Aura::HandleShapeshiftBoosts(bool apply)
                     if ((*i)->GetSpellProto()->SpellIconID == 240 && (*i)->GetModifier()->m_miscvalue == 3)
                     {
                         int32 HotWMod = (*i)->GetModifier()->m_amount;
-                        if(GetModifier()->m_miscvalue == FORM_CAT)
-                            HotWMod /= 2;
 
                         target->CastCustomSpell(target, HotWSpellId, &HotWMod, NULL, NULL, true, NULL, this);
+                        HotWMod /= 2;
+
+                        if (form == FORM_CAT)
+                          {
+							  switch (HotWMod)
+							  {
+								case 2: HotWSpellId = 30902; break;
+								case 4: HotWSpellId = 30903; break;
+								case 6: HotWSpellId = 30904; break;
+								case 8: HotWSpellId = 30905; break;
+								case 10: HotWSpellId = 30906; break;
+							  }
+                          }
+                        else
+                          {
+							  switch (HotWMod)
+							  {
+								case 2: HotWSpellId = 19255; break;
+								case 4: HotWSpellId = 19256; break;
+								case 6: HotWSpellId = 19257; break;
+								case 8: HotWSpellId = 19258; break;
+								case 10: HotWSpellId = 19259; break;
+							  }
+                          }
+
+                        target->CastCustomSpell(target, HotWSpellId, NULL, NULL, NULL, true, NULL, this);
+
                         break;
                     }
                 }
@@ -5978,6 +6000,41 @@ void Aura::HandleShapeshiftBoosts(bool apply)
     }
     else
     {
+        // Heart of the Wild (delete aura)
+        if (form == FORM_CAT || form == FORM_BEAR || form == FORM_DIREBEAR)
+        {
+             Unit::AuraList const& mModTotalStatPct = target->GetAurasByType(SPELL_AURA_MOD_TOTAL_STAT_PERCENTAGE);
+             for(Unit::AuraList::const_iterator i = mModTotalStatPct.begin(); i != mModTotalStatPct.end(); ++i)
+             {
+                 if ((*i)->GetSpellProto()->SpellIconID == 240 && (*i)->GetModifier()->m_miscvalue == 3)
+                 {
+                     int32 HotWMod = (*i)->GetModifier()->m_amount / 2;
+                     if (form == FORM_CAT)
+                     {
+						switch (HotWMod)
+						  {
+							case 2: HotWSpellId = 30902; break;
+							case 4: HotWSpellId = 30903; break;
+							case 6: HotWSpellId = 30904; break;
+							case 8: HotWSpellId = 30905; break;
+							case 10: HotWSpellId = 30906; break;
+						  }
+                     }
+                     else
+                     {
+						switch (HotWMod)
+						  {
+							case 2: HotWSpellId = 19255; break;
+							case 4: HotWSpellId = 19256; break;
+							case 6: HotWSpellId = 19257; break;
+							case 8: HotWSpellId = 19258; break;
+							case 10: HotWSpellId = 19259; break;
+						  }
+                     }
+                     target->RemoveAurasDueToSpell(HotWSpellId);
+                 }
+             }
+        }
         if(spellId1)
             target->RemoveAurasDueToSpell(spellId1);
         if(spellId2)
